### PR TITLE
Disable tracing for specific URIs

### PR DESCRIPTION
**Pull request by SammyK**
### Description

Tracing for specific URI's can be disabled with a comma-separated list of URI's that support `*` as a wildcard. The list of URI's is supplied by the `DD_DISABLE_URI` env var. For example:

`DD_DISABLE_URI=/foo,/users/*,/index.php?foo=*`

To keep things simple and less bug-prone, I decided to add this functionality in the "proceduraly" part of the tracer to ensure this can be checked before any autoloading/detection/injection occurs.

### Readiness checklist
- [x] [Changelog entry](docs/changelog.md) added, if necessary
- [x] Tests added for this feature/bug

### CODE REVIEW

File: bridge/functions.php [hunk 1]
```diff
@@ -10,12 +10,38 @@ function dd_tracing_enabled()
...
+        $requestUri = isset($_SERVER['REQUEST_URI']) ? $_SERVER['REQUEST_URI'] : null;
+    }
+    if (!$requestUri) {
         return true;
     }
+
+    $uris = explode(',', $value);
+    foreach ($uris as $uri) {
+        $uriRegex = str_replace('\*', '.*', preg_quote(trim($uri), '~'));
+        if (1 === preg_match('~^' . $uriRegex . '$~', $requestUri)) {
```
[the remark this answers is not in the document]
Comment (SammyK) [reply bridge/functions.php]: Yeah, it doesn't really support scenarios where you have a lot of different query params that can  be applied in any order. I can remove the query param test from the tests to show we don't support it officially. :)

File: tests/Unit/FunctionsTest.php [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+        public function urisDataProvider()
+        {
+            return [
+                ['', true],
+                ['/', true],
+                ['/foo', false],
+                ['/users', true],
+                ['/users/' . mt_rand(), false],
+                ['/bar/test', true],
+                ['/bar/' . mt_rand() . '/more/test', false],
```
[the remark this answers is not in the document]
Comment (SammyK) [reply tests/Unit/FunctionsTest.php]: I was originally thinking that `*` was greedy, but I can change it to, "matches everything except slashes," if that's more standard/expected. :)

File: bridge/functions.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (labbati) [reply bridge/functions.php]: Do you think then that it would be a good idea to remove the query string parameter alltogether? something like using as the value the result of `strtok($_SERVER['REQUEST_URI'], '?')`. We can always add it back after talking to the various teams.

File: tests/Unit/FunctionsTest.php (hunk 2 above)
[the remark this answers is not in the document]
Comment (labbati) [reply tests/Unit/FunctionsTest.php]: This is a good point. What I am concerned about is that if we include `/` in `*` there will be edge cases that we are not able to cover. While if we exclude `/` from `*` it may be a little bit more cases to write, but they will be able to cover (almost) every use case.

Again I am almost nit-picking here, so feel free to apply your best judgement 😄

**Comment by SammyK** [comment]
Thanks for the excellent CR @labbati! I applied your suggestions and also added the route check to the `Bootstrap::tracerOnce()` method. I know that means that the routes get twice for auto-instrumentation, but I figured this would work until we get filters added and we can remove all this stuff. Does that work for you? :)

File: bridge/functions.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (SammyK) [reply bridge/functions.php]: I love that idea! Done. :)

**Comment by SammyK** [comment]
Closing this outdated PR now that #442 will give use normalized resource names and we can ignore specific URL's at the Agent config level [by setting `DD_APM_IGNORE_RESOURCES`](https://docs.datadoghq.com/agent/docker/apm/?tab=java#docker-apm-agent-environment-variables), e.g. `DD_APM_IGNORE_RESOURCES="GET /health-check"`.
